### PR TITLE
Add Whitening

### PR DESCRIPTION
**Pull request by qbarthelemy**
On second thoughts, the whitening and dimension reduction applied in `AJDC.fit()`, and added in #120, can be useful for many other covariance-based pipelines. @ygerf @sylvchev

This PR:

- adds a new class `Whitening` in `spatialfilters`, with an optional dimension reduction (3 different ways), with covariance matrices as inputs;

- updates `AJDC` to use the new whitening, and replaces parameter `expl_var` by `dim_red`: this replacement is not backward-compatible, but it should not be a problem because it was merged yesterday, and not published in a release (so, this PR should be merged before #121).

### CODE REVIEW

File: pyriemann/spatialfilters.py [hunk 1]
```diff
@@ -1,17 +1,203 @@
...
+            elif self._dr_key == 'max_cond':
+                if self._dr_val <= 1:
+                    raise ValueError(
+                        'Value max_cond must be strictly superior to 1 '
+                        '(Got %d)' % self._dr_val)
+            else:
+                raise ValueError(
+                    'Unknown key in parameter dim_red: %r' % self._dr_key)
+        else:
+            raise ValueError('Unknown type for parameter dim_red')
```
Comment (agramfort) [inline pyriemann/spatialfilters.py:102]: all this parameter validation should be done in fit and not in the init. This violates the scikit-learn API.
The init should not add any other attribute that the params passed to the init.
If someone uses set_params from BaseEstimator after the init you will have a problem otherwise.

File: pyriemann/spatialfilters.py [hunk 2]
```diff
@@ -549,17 +736,15 @@ class AJDC(BaseEstimator, TransformerMixin):
     """
 
     def __init__(self, window=128, overlap=0.5, fmin=None, fmax=None, fs=None,
-                 expl_var=0.999, verbose=True):
+                 dim_red={'max_cond': 100}, verbose=True):
```
Comment (agramfort) [inline pyriemann/spatialfilters.py:739]: I don't like this. To me you should be using a scikit-learn pipeline and not do any preprocessing in the init. It also makes the API more complex with dict which should not be a default parameter as it's a mutable object.

**Review by agramfort — COMMENTED** [review]
would it make sense to move Whitening in a preprocessing submodule rather than in spatialfilters ?

File: pyriemann/spatialfilters.py (hunk 1 above)
Comment (qbarthelemy) [reply pyriemann/spatialfilters.py:102]: You are right: I will test parameters in `fit` rather than in `_init_`.

File: pyriemann/spatialfilters.py (hunk 2 above)
Comment (qbarthelemy) [reply pyriemann/spatialfilters.py:739]: What do you suggest? Defining `None` as default value for `dim_red`?

Comment (agramfort) [reply pyriemann/spatialfilters.py]: I would not offer any dim red option in AJDC and tell users to use a pipeline with one step that does the dim red

Comment (qbarthelemy) [reply pyriemann/spatialfilters.py]: I don't see how to use a `sklearn` pipeline inside `AJDC`.
Because dimension reduction is an inner step of `AJDC.fit()`, applied between cospectra estimation and ajd_pham.

Can you point an example using the suggested architecture?

**Comment by agramfort** [comment]
see eg how we use make_pipeline in :

https://pyriemann.readthedocs.io/en/latest/auto_examples/ERP/plot_classify_MEG_mdm.html

**Comment by qbarthelemy** [comment]
Yes, I know sklearn pipelines. But, once again, cospectra dimension reduction is an inner step inside AJDC.fit(), applied between cospectra estimation and Pham's AJD. This nested sub-step is not at the same level as AJDC, and consequently, it can't be pipelined before or after fitting AJDC.

**Comment by agramfort** [comment]
would it make sense to have a pipeline with cospectra estimation >
whitening > AJDC?

if not you need to explain to me on a call it would be easier

>

**Review by agramfort — COMMENTED** [review]
I'll try to push directly to this branch @qbarthelemy

**Comment by agramfort** [comment]
Great team work @qbarthelemy 💪

**Comment by qbarthelemy** [comment]
Yep! 

Just a minor mistake: new classes `AJDC` and `Whitening` are declared in version `0.2.8`, but it should be `0.2.7` (the actual next release).

**Comment by agramfort** [comment]
good catch

fixed in master

>
